src/mail-app/mail/model/ConversationListModel.ts

```diff
@@ -492,37 +472,73 @@ export class ConversationListModel implements MailSetListModel {
 		return applyInboxRulesToEntries(entries, this.mailSet, this.mailModel, this.inboxRuleHandler)
 	}
 
-	private filterAndUpdateMails(mails: LoadedMail[]): {
+	private addNewLoadedMails(mails: Iterable<LoadedMail>): {
 		addedItems: LoadedConversation[]
-		deletedItems: LoadedConversation[]
 	} {
 		const addedItems: LoadedConversation[] = []
-		const deletedItems: LoadedConversation[] = []
 
 		// store all mails to be loaded later
 		this._updateMails(mails)
 
 		for (const mail of mails) {
-			const conversation = {
-				conversationId: listIdPart(mail.mail.conversationEntry),
-				latestMail: elementIdPart(mail.mailSetEntryId),
-			}
+			const mailSetEntryElementId = elementIdPart(mail.mailSetEntryId)
+			const conversationId = listIdPart(mail.mail.conversationEntry)
+			const existingConversation = this.conversationMap.get(conversationId)
+
+			if (existingConversation == null) {
+				const conversation = {
+					conversationId,
+					latestMail: mailSetEntryElementId,
+					mails: new Set([mailSetEntryElementId]),
+					displayedMail: null,
+				}
+				this.recalculateDisplayedMailForConversation(conversation)
+				addedItems.push(conversation)
+				this.conversationMap.set(conversationId, conversation)
+			} else {
+				existingConversation.mails.add(mailSetEntryElementId)
+				this.recalculateDisplayedMailForConversation(existingConversation)
 
-			const existingConversation = this.conversationMap.get(conversation.conversationId)
-			if (existingConversation != null && this.reverseSortConversation(conversation, existingConversation) >= 0) {
-				continue
+				if (existingConversation.displayedMail === mailSetEntryElementId) {
+					this.listModel.updateLoadedItem(existingConversation)
+				}
 			}
+		}
 
-			if (existingConversation != null) {
-				deletedItems.push(existingConversation)
-				remove(addedItems, existingConversation)
-			}
+		return { addedItems }
+	}
 
-			this.conversationMap.set(conversation.conversationId, conversation)
-			addedItems.push(conversation)
+	/**
+	 * Update the displayed mail to the latest one that matches the current filter, if any.
+	 *
+	 * This does NOT cause a list update.
+	 *
+	 * @param conversation
+	 * @returns true if the displayed mail was changed
+	 * @private
+	 */
+	private recalculateDisplayedMailForConversation(conversation: LoadedConversation): boolean {
+		const currentMail = this.latestMailOfIterable(conversation.mails, this.currentFilter)
+		if (conversation.displayedMail != currentMail) {
+			conversation.displayedMail = currentMail
+			return true
+		} else {
+			return false
 		}
+	}
 
-		return { addedItems, deletedItems }
+	private latestMailOfIterable(mails: Iterable<Id>, filter: ListFilter<Mail> | null): Id | null {
```

Comment:
I was a bit confused with this one, maybe we could name it something like `latestMatchingEmail()`?

I understand now why you do both filtering and finding the latest one in a single pass but it took me a while to understand. I wonder if it's better to keep mails as a sorted array in the `LoadedConversation` already? I can't tell what's more efficient.

I'm also biased, I would from the beginning have put mails there and would just keep an index from mail id to conversation.